### ISSUE TITLE
Is there an option for shard aware?

### ISSUE THREAD

**Issue by pveentjer** [scylladb/scylla-bench#70]
I was wondering why the throughput numbers were so low compared to cassandra-stress (from Scylla) and the number of connections is way lower. So by default scylla-bench isn't shard aware, it doesn't open a connection to every shard and route traffic to the right shard.

**Comment by piodul** [comment]
Which host selection policy are you using (the `host-selection-policy` option)? Shard-awareness should work only with the default option, which is `token-aware`. In that case, the gocql driver should open exactly one connection for each shard and will ignore the `connection-count` parameter. See [here](https://github.com/scylladb/gocql#configuration) for more details.

**Comment by pveentjer** [comment]
I'm using default.

**Comment by piodul** [comment]
How did you build scylla-bench? Like this?

    go install github.com/scylladb/scylla-bench

I just checked and this version indeed is not shard-aware. However, if you build scylla-bench from source it should be shard-aware:

    git clone https://github.com/scylladb/scylla-bench
    go install .

Please check if the version built from source works better.

**Comment by pveentjer** [comment]
I'm using the first approach.

When using '-host-selection-policy=token-aware' the connection count setting is not ignored btw. I set it to 92 and every shard has multiple connections (it seems 3 or 4 per shard).

**Comment by piodul** [comment]
> I'm using the first approach.
> 
> When using '-host-selection-policy=token-aware' the connection count setting is not ignored btw. I set it to 92 and every shard has multiple connections (it seems 3 or 4 per shard).

I'm not exactly sure what is happening here, but I suspect that `go install` builds scylla-bench using the original gocql/gocql driver, not the scylladb/gocql fork. Only the scylladb/gocql fork supports shard-awareness, gocql/gocql does not.

**Comment by pveentjer** [comment]
It seems like a bug :) But let me verify first with the custom build approach.

**Comment by pveentjer** [comment]
I have remove the connection count setting and I'm getting many connections. it seems there are 4/5 connections per shard. This seems pretty high. 

I see there are just a few cross shard operations. So it seems to be shard aware.

So apart from the multiple connections per shard, it seems to be working correctly using the custom build approach.

**Comment by piodul** [comment]
I did some digging and it seems that my theory about `go install` not using our fork is correct. We are using a [replace](https://github.com/scylladb/scylla-bench/blob/d1644082c188494a031eddb710ecf588b3913f44/go.mod#L12) directive to substitute our fork in place of the upstream (which is a recommended practice according to our fork's [readme](https://github.com/scylladb/gocql/blob/35bdbf71260497c5aef7c2d35d02ca6d5c94398f/README.md#installation)). However, commands like `go get` and `go install` do not honor the `replace` directive - it is expected behavior: https://github.com/golang/go/issues/30354 - which resulted in you accidentally using the non-shard-aware driver.

I don't think this issue should classify as a bug in scylla-bench, because the problem was caused by the combination of `go` tool's idiosyncrasies and the way our fork is supposed to be substituted. However, the proper way to install scylla-bench should be documented in the readme - now, the recommended way is `go get`, which is not what we want! I'll send a PR with an improved instruction.
